nullaway/src/main/java/com/uber/nullaway/GenericsChecks.java

```diff
@@ -507,6 +542,207 @@ public void compareGenericTypeParameterNullabilityForCall(
+   *         }
+   *     }
+   * </pre>
+   *
+   * Within the context of class {@code C}, the method {@code Fn.apply} has a parameter type of
+   * {@code @Nullable String}, since {@code @Nullable String} is passed as the type parameter for
+   * {@code P}. Hence, overriding method {@code C.apply} must take a {@code @Nullable String} as a
+   * parameter.
+   *
+   * @param parameterIndex index of the parameter
+   * @param method the generic method
+   * @param enclosingType the enclosing type in which we want to know {@code method}'s parameter
+   *     type nullability
+   * @return nullability of the relevant parameter type of {@code method} in the context of {@code
+   *     enclosingType}
+   */
+  public Nullness getGenericMethodParameterNullness(
+      int parameterIndex, Symbol.MethodSymbol method, Type enclosingType) {
+    Type methodType = state.getTypes().memberType(enclosingType, method);
+    Type paramType = methodType.getParameterTypes().get(parameterIndex);
+    return getTypeNullness(paramType, config);
+  }
+
+  /**
+   * This method compares the type parameter annotations for overriding method parameters with
+   * corresponding type parameters for the overridden method and reports an error if they don't
+   * match
+   *
+   * @param tree tree for overriding method
+   * @param overriddenMethodType type of the overridden method
+   */
+  private void checkTypeParameterNullnessForOverridingMethodParameterType(
+      MethodTree tree, Type overriddenMethodType) {
+    List<? extends VariableTree> methodParameters = tree.getParameters();
+    List<Type> typeParameterTypes = overriddenMethodType.getParameterTypes();
+    for (int i = 0; i < methodParameters.size(); i++) {
+      Type methodParameterType = ASTHelpers.getType(methodParameters.get(i));
+      Type typeParameterType = typeParameterTypes.get(i);
+      if (typeParameterType instanceof Type.ClassType
+          && methodParameterType instanceof Type.ClassType) {
+        if (!compareNullabilityAnnotations(
+            (Type.ClassType) typeParameterType, (Type.ClassType) methodParameterType)) {
+          reportInvalidOverridingMethodParamTypeError(
+              methodParameters.get(i), typeParameterType, methodParameterType);
+        }
+      }
+    }
+  }
+
+  /**
+   * This method compares the type parameter annotations for an overriding method's return type with
+   * corresponding type parameters for the overridden method and reports an error if they don't
+   * match
+   *
+   * @param tree tree for overriding method
+   * @param overriddenMethodType type of the overridden method
+   */
+  private void checkTypeParameterNullnessForOverridingMethodReturnType(
+      MethodTree tree, Type overriddenMethodType) {
+    Type typeParamType = overriddenMethodType.getReturnType();
+    Type methodReturnType = ASTHelpers.getType(tree.getReturnType());
+    if (!(typeParamType instanceof Type.ClassType && methodReturnType instanceof Type.ClassType)) {
```

Comment:
> Do we need a preconditions check here that whenever one of the two is `Type.ClassType`, the other must be as well? Or is there a case where that doesn't hold?

Not sure I fully follow.  Are you asking if we should just check that one of the types is a `ClassType`, and then assert that the other one is also, since they should always be the same?

> Also, presumably arrays of generics is a case to be handled in a future PR, correct? (i.e. returning R[] from a version of apply)

Yes, this is a known issue we plan to handle in the near-ish future.

Reply:
> Not sure I fully follow. Are you asking if we should just check that one of the types is a `ClassType`, and then assert that the other one is also, since they should always be the same?

Yes, unless that's not safe for some reason.

Reply:
Fixed in 08396894725bfbf455bfcc4d1c846c67fc7e3b60